skword2vec/streams.py

```diff
@@ -0,0 +1,196 @@
+            self.limit = limit
+            # functools.update_wrapper(self, gen_func)
+
+        def __iter__(self):
+            if self.limit is not None:
+                return islice(
+                    gen_func(*self.__args, **self.__kwargs), self.limit
+                )
+            return gen_func(*self.__args, **self.__kwargs)
+
+    return _multigen
+
+
+U = TypeVar("U")
+
+
+@reusable
+def chunk(
+    iterable: Iterable[U], chunk_size: int, sample_size: Optional[int] = None
+) -> Iterable[List[U]]:
+    """
+    Generator function that chunks an iterable for you.
+
+    Parameters
+    ----------
+    iterable: Iterable of T
+        The iterable you'd like to chunk.
+    chunk_size: int
+        The size of chunks you would like to get back
+    sample_size: int or None, default None
+        If specified the yielded lists will be randomly sampled with the buffer
+        with replacement. Sample size determines how big you want
+        those lists to be.
+
+    Yields
+    ------
+    buffer: list of T
+        sample_size or chunk_size sized lists chunked from
+        the original iterable.
+    """
+    buffer = []
+    for index, elem in enumerate(iterable):
+        buffer.append(elem)
+        if (index % chunk_size == (chunk_size - 1)) and (index != 0):
+            if sample_size is None:
+                yield buffer
+            else:
+                yield random.choices(buffer, k=sample_size)
+            buffer = []
+
+
+@reusable
+def stream_files(
+    paths: Iterable[str],
+    lines: bool = False,
+    not_found_action: Literal["exception", "none", "drop"] = "exception",
+) -> Iterable[Optional[str]]:
+    """Streams text contents from files on disk.
+
+    Parameters
+    ----------
+    paths: iterable of str
+        Iterable of file paths on disk.
+    lines: bool, default False
+        Indicates whether you want to get a stream over lines
+        or file contents.
+    not_found_action: {'exception', 'none', 'drop'}, default 'exception'
+        Indicates what should happen if a file was not found.
+        'exception' propagates the exception to top level, 'none' yields
+        None for each file that fails, 'drop' ignores them completely.
+
+    Yields
+    ------
+    str or None
+        File contents or lines in files if lines is True.
+        Can only yield None if not_found_action is 'none'.
+    """
+    for path in paths:
+        try:
+            with open(path) as in_file:
+                if lines:
+                    for line in in_file:
+                        yield line
+                else:
+                    yield in_file.read()
+        except FileNotFoundError as e:
+            if not_found_action == "exception":
+                raise FileNotFoundError(
+                    f"Streaming failed as file {path} could not be found"
+                ) from e
+            elif not_found_action == "none":
+                yield None
+            else:
```

Comment:
if you make a typo in the literal, the behavior will default to `"drop"`.
it's an easy mistake to make & it could take you hours figuring out what's not working.

Reply:
Spitting straight facts, will address.